File: lib/actions/lists.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';
import 'package:figstyle/state/user.dart';
import 'package:figstyle/types/quote.dart';
import 'package:figstyle/types/user_quotes_list.dart';

Future<UserQuotesList> createList({
  @required BuildContext context,
  String name = '',
  String description = '',
  String iconUrl = '',
  bool isPublic = false,
}) async {
  try {
    final userAuth = stateUser.userAuth;

    final docRef = await FirebaseFirestore.instance
        .collection('users')
        .doc(userAuth.uid)
        .collection('lists')
        .add({
      'createdAt': DateTime.now(),
      'description': description,
      'name': name,
      'itemsCount': 0,
      'iconUrl': iconUrl,
      'isPublic': isPublic,
      'updatedAt': DateTime.now(),
    });

    final doc = await docRef.get();

    final data = doc.data();
    data['id'] = doc.id;

    return UserQuotesList.fromJSON(data);
  } catch (error) {
    debugPrint(error.toString());
    return null;
  }
}

Future<bool> deleteList({
  @required BuildContext context,
  @required String id,
}) async {
  try {
    final userAuth = stateUser.userAuth;

    // Add a new document containing information
    // to delete the subcollection (in order to delete its documents).
    await FirebaseFirestore.instance.collection('todelete').add({
      'objectId': id,
      'path': 'users/<userId>/lists/<listId>/quotes',
      'userId': userAuth.uid,
      'target': 'list',
      'type': 'subcollection',
    });

    // Delete the quote collection doc.
    await FirebaseFirestore.instance
        .collection('users')
        .doc(userAuth.uid)
        .collection('lists')
        .doc(id)
        .delete();

    return true;
  } catch (error) {
    debugPrint(error.toString());
    return false;
  }
}

Future<bool> removeFromList({
  @required BuildContext context,
  @required String id,
  @required Quote quote,
}) async {
  try {
    final userAuth = stateUser.userAuth;

    await FirebaseFirestore.instance
        .collection('users')
        .doc(userAuth.uid)
        .collection('lists')
        .doc(id)
        .collection('quotes')
        .doc(quote.id)
        .delete();

    return true;
  } catch (error) {
    debugPrint(error.toString());
    return false;
  }
}

Future<bool> updateList({
  @required BuildContext context,
  String id = '',
  String iconUrl = '',
  String name = '',
  String description = '',
  bool isPublic = false,
}) async {
  try {
    final userAuth = stateUser.userAuth;

    await FirebaseFirestore.instance
        .collection('users')
        .doc(userAuth.uid)
        .collection('lists')
        .doc(id)
        .update({
      'description': description,
      'name': name,
      'iconUrl': iconUrl,
      'isPublic': isPublic,
      'updatedAt': DateTime.now(),
    });

    return true;
  } catch (error) {
    debugPrint(error.toString());
    return false;
  }
}

File: functions/src/lists.ts
import * as functions from 'firebase-functions';
import { adminApp } from './adminApp';
import { checkUserIsSignedIn } from './utils';

const firestore = adminApp.firestore();

/**
 * Delete an user's quotes list.
 * Add a new document to `todelete` collection 
 * as the quote sub-collection `list/quotes/{quoteId}` will be delete later.
 */
export const deleteList = functions
  .region('europe-west3')
  .https
  .onCall(async (data: DeleteListParams, context) => {
    const userAuth = context.auth;
    const { listId, idToken } = data;

    if (!userAuth) {
      throw new functions.https.HttpsError(
        'unauthenticated', 
        `The function must be called from an authenticated user.`,
      );
    }

    await checkUserIsSignedIn(context, idToken);

    if (!listId) {
      throw new functions.https.HttpsError(
        'invalid-argument', 
        `The function must be called with one argument "listId" which is the list to delete.`
      );
    }

    const listSnapshot = await firestore
      .collection('users')
      .doc(userAuth.uid)
      .collection('lists')
      .doc(listId)
      .get();


    const listData = listSnapshot.data();

    if (!listSnapshot.exists || !listData) {
      return {
        success: false,
        error: {
          message: "This list doesn't exist anymore.",
        },
        uid: userAuth.uid,
        target: {
          type: 'list',
          id: listId,
          date: Date.now(),
        }
      }
    }

    await firestore
      .collection('todelete')
      .doc(listId)
      .set({
        doc: {
          id: listId,
          conceptualType: 'list',
          dataType: 'subcollection',
          hasChildren: true,
        },
        path: `users/${userAuth.uid}/lists/${listId}/quotes`,
        task: {
          createdAt: Date.now(),
          done: false,
          items: {
            deleted: 0,
            total: listData.itemsCount ?? 0,
          },
          updatedAt: Date.now(),
        },
        user: {
          id: userAuth.uid,
        },
      });

    await firestore
      .collection('users')
      .doc(userAuth.uid)
      .collection('lists')
      .doc(listId)
      .delete();

    return {
      user: {
        id: userAuth.uid,
      },
      target: {
        type: 'list',
        id: listId,
        date: Date.now(),
      }
    }
  });

/**
 * Increment user's list count on create list.
 */
export const onListAdded = functions
  .region('europe-west3')
  .firestore
  .document('users/{userId}/lists/{listId}')
  .onCreate(async ({}, context) => {
    const user = await firestore
      .collection('users')
      .doc(context.params.userId)
      .get();

    if (!user.exists) { return; }
    
    const userData = user.data();
    if (!userData) { return; }

    const userLists: number = userData.stats.lists ?? 0;
    
    return await user.ref
      .update('stats.lists', userLists + 1);
  });

/**
 * Decrement user's list count on create list.
 */
export const onListDeleted = functions
  .region('europe-west3')
  .firestore
  .document('users/{userId}/lists/{listId}')
  .onDelete(async ({}, context) => {
    const user = await firestore
      .collection('users')
      .doc(context.params.userId)
      .get();

    if (!user.exists) { return; }

    const userData = user.data();
    if (!userData) { return; }

    const userLists: number = userData.stats.lists ?? 0;

    return await user.ref
      .update('stats.lists', Math.max(0, userLists - 1));
  });

/**
 * Increment user's list `itemsCount` when a new quote is added.
 */
export const onQuoteAdded = functions
  .region('europe-west3')
  .firestore
  .document('users/{userId}/lists/{listId}/quotes/{quoteId}')
  .onCreate(async ({}, context) => {
    const { userId, listId } = context.params;

    const listSnapshot = await firestore
      .collection('users')
      .doc(userId)
      .collection('lists')
      .doc(listId)
      .get();

    const listData = listSnapshot.data();

    if (!listSnapshot.exists || !listData) {
      return;
    }

    const itemsCount = listData.itemsCount ?? 0;

    await listSnapshot
      .ref
      .update({
        itemsCount: itemsCount + 1,
      })
  });

/**
 * Decrement user's list `itemsCount` when a new quote is added.
 */
export const onQuoteDeleted = functions
  .region('europe-west3')
  .firestore
  .document('users/{userId}/lists/{listId}/quotes/{quoteId}')
  .onDelete(async ({}, context) => {
    const { userId, listId } = context.params;

    const listSnapshot = await firestore
      .collection('users')
      .doc(userId)
      .collection('lists')
      .doc(listId)
      .get();

    const listData = listSnapshot.data();

    if (!listSnapshot.exists || !listData) {
      return;
    }

    const itemsCount = listData.itemsCount ?? 0;

    await listSnapshot
      .ref
      .update({
        itemsCount: itemsCount - 1,
      })
  });
  